### ISSUE TITLE
NaN problem of dcrec_seq model

### ISSUE THREAD

**Issue by KNwbq** [HKUDS/SSLRec#11]
Firstly, I am grateful for your valuable contribution to applying self-supervised learning in the recommendation system. 

### Problem Description
I discovered that the `kl_loss` of the DCRec model is yielding a value of `nan` during the training of the DCRec_seq model.
### Steps to reproduce
1. Run the code by the `python main.py --model dcrec_seq`
2. Check out the log of loss, which is `[Epoch  49 /  50] loss: 5.5758 cl_loss: 0.0003 kl_loss: nan`  at the last epoch
### Problem Cause
Upon delving into the code, I found the main reason for the problem is the following code:
```
expected_weights_distribution = torch.normal(
            self.weight_mean, 0.1, size=mainstream_weights.size()).sort()[0].to(self.device)
        kl_loss = self.kl_weight * F.kl_div(F.log_softmax(
            mainstream_weights, dim=0).sort()[0], expected_weights_distribution, reduction="batchmean")
```
the `expected_weights_distribution` is a normal distribution with the mean of `self.weight_mean` and the std of `0.1`. However, we should notice that the value of the `expected_weights_distribution`, obtained by `torch.normal`, may be non-positive, which will cause the nan problem during the calculation of KL divergence.
### Problem Solution
Due to the problem cause, I modify the code as follows:
```
        expected_weights_distribution = torch.normal(
            self.weight_mean, 0.1, size=mainstream_weights.size()).sort()[0].to(self.device)
        kl_loss = self.kl_weight * F.kl_div(F.log_softmax(
            mainstream_weights, dim=0).sort()[0], F.softmax(expected_weights_distribution, dim=-1), reduction="batchmean")
```
I simply apply the `softmax` function to ensure the value of `expected_weights_distribution` is positive. However, I have no idea whether the distribution of the handled one is still normal.

I would appreciate your reply.

**Comment by yuh-yang** [comment]
Hi,

Thanks for issuing this bug. This NaN bug caused by KL Divergence calculation in DCRec was suspected and fixed in the individual repository of the paper: [link](https://github.com/HKUDS/DCRec/commit/23382332f79351b1340033946982b4dd93ae46ae#diff-e6fbc50b8064f0b23b5377ccd8d0bf44d8f7950f35afe8017c0441eee54dd403)

And you are right the reason is that torch.normal can have non-positive values for log(). A simple workaround is to truncate these values to a tiny number (like 1e-8):

```
def cal_kl_1(target, input):
    target[target<1e-8] = 1e-8
    target = torch.log(target + 1e-8)
    input = torch.log_softmax(input + 1e-8, dim=0)
    return F.kl_div(input, target, reduction='batchmean', log_target=True)
```

I tested this and it works well without a noticeable performance decline. I'm not quite sure if applying softmax is another effective workaround. However, I have now updated this bugfix for DCRec here, and please feel free to test and compare. :)

**Comment by KNwbq** [comment]
Got it! Thank you for your reply.
